R/bootstrap.R

```diff
@@ -43,19 +43,27 @@ bootstrapPage <- function(..., title = NULL, responsive = deprecated(), theme =
   }
 
   ui <- tagList(
-    bootstrapLib(theme),
     if (!is.null(title)) tags$head(tags$title(title)),
-    # TODO: throw better error when length > 1?
     if (is.character(theme)) {
-      tags$head(tags$link(rel="stylesheet", type="text/css", href = theme))
+      if (length(theme) > 1) stop("`theme` must point to a single CSS file, not multiple files.")
+      tags$head(tags$link(rel="stylesheet", type="text/css", href=theme))
     },
     # remainder of tags passed to the function
     list(...)
   )
 
-  ui <- setLang(ui, lang)
+  # If theme is a bslib::bs_theme() object, bootstrapLib() needs to come first
+  # (so other tags, when rendered via tagFunction(), know about the relevant
+  # theme). However, if theme is anything else, we intentionally avoid changing
+  # the tagList() contents to avoid breaking user code that makes assumptions
+  # about the return value https://github.com/rstudio/shiny/issues/3235
+  if (is_bs_theme(theme)) {
+    ui <- do.call(tagList, c(bootstrapLib(theme), ui))
```

Comment:
I started writing a comment explaining this potentially-confusing usage of a tagList as a regular list (of arguments), and how it could be changed back to a normal function call (without `do.call`) in the future if we decide to get rid of the `else` code path, but it was getting long and I think it's just easier to write the code the way I did above. (My goal is to avoid situations where people are afraid to touch the code because they don’t understand what it does.)